### PR TITLE
CMake: improve build scripts

### PR DESCRIPTION
**Pull request by oltolm**
# Description

Improve CMake build scripts. WolfSSL is used by rpcs3, so some problems with the CMake build scripts became apparent.

* WolfSSL needs the following `target_compile_definitions(wolfssl INTERFACE SIZEOF_LONG=${SIZEOF_LONG} SIZEOF_LONG_LONG=${SIZEOF_LONG_LONG})` to compile with Mingw-w64 if `WOLFSSL_FAST_MATH=ON`. I fixed it in `types.h`.
* Linking with WolfSSL fails because it needs to link against `crypt32.lib`.
* The default for `WOLFSSL_BUILD_OUT_OF_TREE` is `OFF`, which means that `config.h` is generated in the source tree and is invisible because it is ignored by git. I changed the default to out-of-tree-build and removed `config.h` from `.gitignore`.
* The build scripts do not export WolfSSL which makes it hard for other projects to use. Now you can use `find_package(WolfSSL)`. To properly export WolfSSL i changed `add_definitions` to `target_compile_definitions`.

# Testing

I compiled WolfSSL as part of rpcs3 and stand-alone.

### CODE REVIEW

**Comment by dgarske** [comment]
Hi @oltolm ,

Thank you for these CMake fixes. In order to accept this we'll need a signed contributor agreement. Can you email support@ wolfssl dot com and mention this PR and provide some additional details about your project and location?

Thanks,
David Garske, wolfSSL

**Comment by oltolm** [comment]
Hi David,

I sent an email to the support.

Am Mo., 24. Apr. 2023 um 18:37 Uhr schrieb David Garske <
***@***.***>:

**Comment by dgarske** [comment]
Signed contributor agreement received and approved.

**Review by dgarske — CHANGES_REQUESTED** [review]
Great fixes here. Just one comment about the location of the new cmake file.

**Comment by dgarske** [comment]
OK to test

File: .gitignore [hunk 1]
```diff
@@ -10,8 +10,9 @@ ctaocrypt/src/src/
 *.cache
 .dirstamp
 *.user
-configure
-config.*
+configure
+config.log
```
Comment (dgarske) [inline .gitignore:14]: Please revert the `config.*` change. There are several files that get generated that need to be ignored. Like `config.h` and `config.in`. Thanks

**Review by dgarske — CHANGES_REQUESTED** [review]
Just one small change then it is ready for merge. Thank you @oltolm

File: .gitignore (hunk 1 above)
Comment (oltolm) [reply .gitignore]: I reverted them, but now you won't notice the generated files, and you can always ignore them using `.git/info/exclude`.

**Review by dgarske — APPROVED** [review]

File: cmake/Config.cmake.in [hunk 2]
```diff
@@ -0,0 +1,3 @@
+@PACKAGE_INIT@
```
Comment (dgarske) [inline cmake/Config.cmake.in:1]: This file needs added to cmake/include.am. Fixes our “make dist” 

added: cmake/Config.cmake.in

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]
Over to @JacobBarthelmeh to finalize. Please squash on merge.

**Comment by dgarske** [comment]
Retest this please

**Review by JacobBarthelmeh — APPROVED** [review]

**Comment by JacobBarthelmeh** [comment]
@oltolm thank you for your work on this! It is a good step in the right direction for cmake + mingw32 support. When building wolfSSL were you able to successfully run the crypto unit tests? Using `./wolfcrypt/test/testwolfcrypt.exe`. When testing locally with the mingw32 + cmake build I was running into some SHA384 failures.

**Comment by oltolm** [comment]
I use the mingw32 and mingw64 [subsystems](https://www.msys2.org/wiki/MSYS2-introduction/#subsystems) of MSYS2. If I just build with default CMake options then the test runs successfully in mingw32 and mingw64.

**Pull request merged**